s3/s3.go

```diff
@@ -192,16 +192,22 @@ func (b *Bucket) Put(path string, data []byte, contType string, perm ACL) error
 // PutReader inserts an object into the S3 bucket by consuming data
 // from r until EOF.
 func (b *Bucket) PutReader(path string, r io.Reader, length int64, contType string, perm ACL) error {
-	headers := map[string][]string{
-		"Content-Length": {strconv.FormatInt(length, 10)},
-		"Content-Type":   {contType},
-		"x-amz-acl":      {string(perm)},
-	}
+	return b.PutReaderWithHeader(path, r, length, contType, perm, http.Header{})
+}
+
+// PutReader inserts an object into the S3 bucket by consuming data
+// from r until EOF. The HTTP PUT request will have the additional
+// headers provided. This is useful for cache control etc.
+func (b *Bucket) PutReaderWithHeader(path string, r io.Reader, length int64, contType string, perm ACL, hdrs http.Header) error {
+	hdrs.Set("Content-Length", strconv.FormatInt(length, 10))
+	hdrs.Set("Content-Type", contType)
+	hdrs.Set("x-amz-acl", string(perm))
```

Comment:
This line is problematic if someone wants to use these headers: x-amz-grant-read, x-amz-grant-write, x-amz-grant-full-control: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html (see **Access Permissions**)

These headers were not possible before this commit either of course...